### ISSUE TITLE
Make emscripten builds running on Travis

### ISSUE THREAD

**Issue by musicEnfanthen** [rism-ch/verovio#1179]
Finally managed to have the built run successfully on Travis: [https://travis-ci.org/musicEnfanthen/verovio/builds/605527164](https://travis-ci.org/musicEnfanthen/verovio/builds/605527164)

It took 1h 33mins and I had to split up the existing `deploy-develop.sh` build script, cf. the ci-prefixed scripts here: [https://github.com/musicEnfanthen/verovio/tree/travis-test](https://github.com/musicEnfanthen/verovio/tree/travis-test)

Here's the corresponding `yaml` file: [https://github.com/musicEnfanthen/verovio/blob/travis-test/.travis.yml](https://github.com/musicEnfanthen/verovio/blob/travis-test/.travis.yml)

Also tried to run the 4 different build jobs in parallel, what went well in general and reduced the total time to about 35-40 mins. But it is not possible to collect the build artifacts from parallel jobs, as the cache will be reinitiated and overwritten for every parallel job (what is not the case in sequential jobs).

So there is still quite a lot to optimize. One idea is to commit every build artifact after any parallel build job has finished and to [squash merge](https://stackoverflow.com/questions/5308816/how-to-use-git-merge-squash) at the end in the deploy stage (so only one rebased commit would be pushed). But I am not sure if the squash merge is possible for parallel builds at all, since the cache will be overwritten, i.e. also the temporary gh-pages.git in travis will be overwritten for every parallel stage, so the commits would probably not be persistent. (At least, I tried to do so with `git add -A` what did not work. Only the file changes of the last build stage (most of the time the build including humdrum) did "survive" to the `deploy` stage.)

Another idea would be to push every build stage as soon as it has succesfully finished. But then one should check if the remote repo has been changed in the meantime (especially by a push from another build).

Another possibility is to upload build artifacts to another server and download them before the deploy. This is documented in Travis with amazon S3: [https://docs.travis-ci.com/user/uploading-artifacts/](https://docs.travis-ci.com/user/uploading-artifacts/) But I guess, that's nothing you actually want for verovio.

Other optimization options:
- as mentioned by @lpugin on slack: use docker prebuild for emscripten ([https://emscripten.org/docs/compiling/Travis.html](https://emscripten.org/docs/compiling/Travis.html))
- check if `emscripten` build time can be optimized (especially the last step in the emscripten build takes most of the build time on Travis, for the build including humdrum it is more than 10 mins only for this final step)
- use docker to "cache" build artifacts?

**Comment by lpugin** [comment]
This is great! many thanks. Would it be possible to simply commit and push each build separately? Since they each modify different files it should not be a problem and not create any conflicts.

**Comment by musicEnfanthen** [comment]
It took a while, but it seems to work now: [https://travis-ci.org/musicEnfanthen/verovio/builds/609007027](https://travis-ci.org/musicEnfanthen/verovio/builds/609007027).

You can find the travis.yml and the ci-scripts here on the `travis-test` branch: [https://github.com/musicEnfanthen/verovio/tree/travis-test](https://github.com/musicEnfanthen/verovio/tree/travis-test)

The total build time was about 50 mins, the builds are running in parallel, and after completion the artifacts are pushed separately to a temporary output branch (that has a unique name according to the hash of the commit that triggered the build). In the deploy stage, all the commits of the temporary branch are squash merged (see above) into `gh-pages` as a single commit, so you don't spam the `gh-pages` branch with at least 8 commits for every build. Finally, after the merge is completed, the temporary branch is safely deleted, and we have a new build commit: [https://github.com/musicEnfanthen/verovio/commit/134e0de24a6cdf015bb7eac43eb4c10661f17fcd](https://github.com/musicEnfanthen/verovio/commit/134e0de24a6cdf015bb7eac43eb4c10661f17fcd)

If this setup would be OK with you, you definitely should have some more test runs on your side. Tried to configure everything with global variables in the `travis.yml`, so switching branch names (to some temporary test branches) should be as easy as changing them in the `travis.yml`.

And probably, in the end, all those build scripts should go into some separate folder to keep the root clean and simple.

**Comment by rettinghaus** [comment]
Extremely cool! 
This allows forks to easily build a special tailored toolkit. Could you please add some hints to the wiki where and how to adapt the build scripts?

**Comment by lpugin** [comment]
@musicEnfanthen would it be possible to group the scripts under a `./travis` directory?

**Comment by lpugin** [comment]
> This allows forks to easily build a special tailored toolkit.

What do you mean? I think people who need this can easily figure out what to change, can't they?

**Comment by rettinghaus** [comment]
Maybe, but wouldn't it be nice to point to the relevant script instead of letting everybody go through all those scripts?

**Comment by musicEnfanthen** [comment]
> @musicEnfanthen would it be possible to group the scripts under a `./travis` directory?

Yes, of course. I moved the scripts to a subdirectory. I'll file a PR on a dedicated branch. Let me know, if you want other changes please. (Later, we should have a look at the new [Github Actions](https://github.com/features/actions). Maybe all of this can be achieved more easily...)

@rettinghaus Actually, there is not much to change. Most important, you would have to adjust the branches from which and on which you will build the toolkits and the repo name accordingly. This can be easily done in the `.travis.yaml` file:

```
branches:
  only:
    - travis-test # TODO: switch back to "develop"

[skip]

env:
  global:
    # repos
    - VEROVIO_REPOSITORY: "https://${GH_TOKEN}@github.com/musicEnfanthen/verovio"  # TODO: switch back to "https://${GH_TOKEN}@github.com/rism-ch/verovio"
    # branches
    - BUILD_BRANCH: travis-test   # TODO: switch back to "develop"
    - GH_PAGES_BRANCH: gh-pages
```

**Comment by lpugin** [comment]
>  I'll file a PR on a dedicated branch. Let me know, if you want other changes please. 

Thanks, please go ahead.

**Comment by musicEnfanthen** [comment]
Should have been automatically closed by #1190
